consensus/scp/tests/mock_network/mod.rs

```diff
@@ -563,17 +512,206 @@ impl SCPNode {
     }
 }
 
-pub fn random_str(rng: &mut StdRng, len: usize) -> String {
-    const CHARSET: &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZ\
-                            abcdefghijklmnopqrstuvwxyz\
-                            0123456789";
+///////////////////////////////////////////////////////////////////////////////
+/// Test Helper
+///////////////////////////////////////////////////////////////////////////////
 
-    let output: String = (0..len)
-        .map(|_| {
-            let idx = (rng.next_u64() % CHARSET.len() as u64) as usize;
-            char::from(CHARSET[idx])
-        })
-        .collect();
+/// Injects values to a network and waits for completion
+pub fn run_test(mut network: SCPNetwork, network_name: &str, options: TestOptions, logger: Logger) {
+    if options.submit_in_parallel {
+        log::info!(
+            logger,
+            "( testing ) begin test for {} with {} values in parallel",
+            network_name,
+            options.values_to_submit,
+        );
+    } else {
+        log::info!(
+            logger,
+            "( testing ) begin test for {} with {} values in sequence",
+            network_name,
+            options.values_to_submit,
+        );
+    }
+
+    let start = Instant::now();
+
+    let mut rng = mc_util_test_helper::get_seeded_rng();
+    let mut values = Vec::<String>::with_capacity(options.values_to_submit);
+    for _i in 0..options.values_to_submit {
+        let value = mc_util_test_helper::random_str(&mut rng, 20);
+        values.push(value);
+    }
+
+    log::info!(
+        network.logger,
+        "( testing ) finished generating {} values",
+        options.values_to_submit
+    );
+
+    let num_nodes: usize = {
+        network
+            .nodes_map
+            .lock()
+            .expect("lock failed on nodes_map getting length")
+            .len()
+    };
+
+    let mut last_log = Instant::now();
+    for i in 0..options.values_to_submit {
+
+        if options.submit_in_parallel {
+            // simulate broadcast of values to all nodes in parallel
+            for n in 0..num_nodes {
+                network.push_value(&test_utils::test_node_id(n as u32), &values[i]);
+            }
+        } else {
+            // submit values to nodes in sequence
+            let n = i % num_nodes;
+            network.push_value(&test_utils::test_node_id(n as u32), &values[i]);
+        }
+
+        std::thread::sleep(Duration::from_micros(
+            1_000_000 / options.submissions_per_sec,
+        ));
+
+        if last_log.elapsed().as_secs() > 1 {
+            log::info!(
+                network.logger,
+                "( testing ) pushed {}/{} values",
+                i,
+                options.values_to_submit
+            );
+            last_log = Instant::now();
+        }
+    }
+
+    // report end of value push
+    log::info!(
+        network.logger,
+        "( testing ) pushed {}/{} values",
+        options.values_to_submit,
```

Comment:
I believe this is incorrect

Reply:
can you elaborate? I'm not sure I understand...

Reply:
The values pushed is not `options.values_to_submit` (the part before the `/` in `{}/{}`)

Reply:
This is still an issue.

Reply:
I think this is the only way the for loop exits, but I'll add a separate check.